:necktie: refactor a bit and add golang ci lint

darkweaver87 opened the pull request:
Thanks for this nice projet :-)

This PR:
- updates code to go 1.19
- adds golangci check
- refactors a bit the code to reduce its complexity
- prepares the ground for another upcoming PR which will separate gitops resources gathering and workload clusters resources listing

darkweaver87 committed:
:necktie: refactor a bit and add golang ci lint

raffis commented on Makefile:
```diff
@@ -23,6 +23,9 @@ fmt:
 test:
 	go test -coverprofile coverage.out -v ./...
 
+lint:
+	golangci-lint run
+
```
This will fail on the ci pipeline as golang-lint would not be found.
You can install it as separate make target as I did in another project: https://github.com/raffis/kjournal/blob/next/Makefile#L59-L66

darkweaver87 replied:
I'm used to CI environments where it's already there 😅 good catch !

raffis commented on cmd/main.go:
```diff
@@ -75,77 +99,76 @@ func init() {
+func detectZombies(logger stderrLogger, flags args, gitopsDynClient, clusterDynClient dynamic.Interface, clusterDiscoveryClient *discovery.DiscoveryClient, clusterRestClient *rest.RESTClient, namespace string) ([]unstructured.Unstructured, error) {
+	var zombies []unstructured.Unstructured
 
+	logger.Debugf("⎈ Helm releases ⎈")
```
A real debug message here would be better, like `discover all helmreleases`

raffis commented on cmd/main.go:
```diff
@@ -75,77 +99,76 @@ func init() {
 	}
 
-	kustomizations, err := listKustomizations(context.TODO(), structClient)
+	logger.Debugf("👷 Kustomizations 👷")
```
Same here, `discover all kustomizations`.

Also I don't have anything against emojis, but I reckon its more professional without them, especially in error types (like used in some other parts of the pr).

raffis commented on cmd/main.go:
```diff
@@ -212,28 +221,106 @@ func run(cmd *cobra.Command, args []string) error {
 	wgConsumer.Add(1)
 	go func() {
 		defer wgConsumer.Done()
-		printer(ch)
+		for res := range ch {
+			zombies = append(zombies, res)
+		}
 	}()
 
 	wgProducer.Wait()
 	close(ch)
 	wgConsumer.Wait()
 
-	return nil
+	return zombies, nil
```
I think dumping results while discovering has its advantages, especially on big clusters. While on the other hand this change enables us to pretty print a nice table.
I would prefer a new argument called `--no-stream`. If given it would cache results like done in this pr. By default results would be dumped immediately.

raffis commented on cmd/main.go:
```diff
@@ -75,77 +99,76 @@ func init() {
+		return nil, err
+	}
+	for _, g := range list {
+		logger.Debugf(g.GroupVersion)
```
A full debug phrase would be nice here instead just dumping the group name.

raffis commented on cmd/main.go:
```diff
@@ -75,77 +99,76 @@ func init() {
+		logger.Debugf(k.GetName())
+	}
+
+	logger.Debugf("👨‍👩‍👧‍👧 Groups 👨‍👩‍👧‍👧")
```
Same here

raffis reviewed (changes requested):
Awesome pr! 🎉 🥇 

I have just some small comments/concerns.

darkweaver87 commented:
I'm fine with all those comments 🫡 I'll push a commit today. Thanks 🙂

darkweaver87 committed:
:beers: apply CR suggestions

- add golangci-lint install
- remove emojis
- add no-steam option + take this opportunity to print a summary
- change exit code

raffis commented on README.md:
```diff
@@ -51,7 +51,7 @@ Flags:
       --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
       --as-uid string                  UID to impersonate for the operation.
-      --cache-dir string               Default cache directory (default "/home/raffi/.kube/cache")
+      --cache-dir string               Default cache directory (default "/home/johndoe/.kube/cache")
```
This should be changed to ~/.kube/cache

darkweaver87 replied:
:+1:

raffis commented on cmd/main.go:
```diff
@@ -29,123 +30,156 @@ const (
+		if err != nil {
+			return err
+		}
+		fmt.Printf("\nSummary: %d resources found, %d zombies detected\n", resourceCount, len(zombies))
```
I like the summary print here, but this should be skipped if the output is set to something other than the default. Otherwise this leads to invalid output for example if set to json.

darkweaver87 replied:
:+1:

raffis commented on cmd/main.go:
```diff
@@ -29,123 +30,156 @@ const (
+		}
+		fmt.Printf("\nSummary: %d resources found, %d zombies detected\n", resourceCount, len(zombies))
+
+		if len(zombies) > 0 {
+			return fmt.Errorf("%d zombies found", len(zombies))
+		}
```
Not sure how to feel about this. exit > 0 implies something did not work which is not the case here. Although I see the need for this therefore it makes sense.

Anyhow returning an error here also ends on stdout which will lead to the same problem as commented above.
Also this would double print the number of zombies twice, first in the summary above, and 2nd in the error returned which is not that nice.

darkweaver87 replied:
In my head, I would use the exit code to send notifications or whatever. What do you think about a `--fail` option like for `curl` ? 
Concerning the code it self you're right I will find another way to achieve  this :-)

raffis replied:
Yes I think a separate argument to achieve this is the better approach 👍🏻

darkweaver87 replied:
OK thanks :-)

raffis reviewed (changes requested):
Thanks for the changes 🚀 , just three concerns left.

darkweaver87 committed:
:beers: apply CR suggestions

- add fail flag
- print summary only with default output

raffis commented on Makefile:
```diff
@@ -12,7 +12,7 @@ endif
 rwildcard=$(foreach d,$(wildcard $(addsuffix *,$(1))),$(call rwildcard,$(d)/,$(2)) $(filter $(subst *,%,$(2)),$(d)))
 
-all: test build
+all: lint test build
```
The lint target is actually not executed during the pr build, can you add that here? https://github.com/raffis/gitops-zombies/blob/b8c6400e39a8aa63cf82cb78ab65aa4f5bcce3a5/.github/workflows/e2e.yaml#L27

darkweaver87 committed:
apply CR suggestions

raffis reviewed (approved):
Thanks for the awesome contribution 🚀 🥇 , I will add a post merge  fix for the lint step as the problem is in the go-install-tool function and unrelated to this pr. For the failing size-label job there is a separate issue #6.
